### DATA SCIENCE AND SUPPLY CHAINS ANALYTICS. A – Z WITH PYTHON

### Sección 18: Seasonal Inventory (Inventario Estacional):

### Intro:

Entonces, hasta ahora, hemos discutido el inventario regular y cuando recibimos el inventario, el mismo producto durante un período de tiempo limitado o durante un período de tiempo regular, pero no para discutir sobre los productos que no son regulares o irregulares, por lo que puedo llamarlos productos de temporada. Por ejemplo, en el comercio minorista, en la industria de la confección, de la ropa duran una temporada, tal vez para sazonar, y luego, en la próxima temporada, las vendemos con un descuento o por un rescate o por un valor de rescate en este tipo de cosas.

Entonces estos son productos típicos de temporada que puedes pensar en las pelotas de la Copa del Mundo. Los balones de la Copa del Mundo duran solo durante la temporada de la Copa del Mundo. Después de esto, se venderá con un gran descuento.

Los difonos se pueden considerar un producto de temporada porque todos los años llega el nuevo modelo y este también se considera un producto de temporada. Entonces, siempre que la venta de un artículo es por un período de tiempo limitado, lo llamamos un producto de temporada y las características del producto de temporada son un poco diferentes a las características del producto regular, porque en las características del producto de temporada, nos preocupamos un poco por las ganancias que vienen. de este producto.

Así que tomemos un pequeño ejemplo. Digamos que soy el fabricante de los balones de la Copa del Mundo y este plato de la Copa del Mundo va a durar. Las ventas de la misma se mantendrán durante la temporada de la Copa del Mundo. Entonces, digamos que cinco minutos me costó hacer el tazón de la Copa del Mundo con mi proveedor, digamos, 40 USD, así que este es el costo del bote y venderé este Centro Comercial de la Copa del Mundo por 65 USD.

Durante la temporada de la Copa del Mundo, entonces, ¿qué perdí, qué perderé si compré 50 tazones de la Copa del Mundo y 51 cliente vinieron y me pidieron este tazón y para el último cliente, no tuve suficiente? stock para él, así que se fue. ¿Qué perdí en este caso?

Bueno, creo que he perdido las ganancias, las ganancias provenientes de este barco. Entonces podemos llamar a esto el costo de la escasez y el costo de la escasez será. 25, está bien, no qué y luego qué perderé si obtuve 52 votos y vinieron 51 clientes. ¿Qué perdí al final?

Bueno, creo que perdí el costo del cuenco en sí. Los 40 dólares por lo que el costo del exceso. Él tiene 40 años, así que lo que está tratando de hacer es ver cuál es la mejor cantidad que podemos obtener para maximizar nuestras ganancias, y usamos algo llamado problema crítico y el problema crítico es simplemente igual a C.

S sobre C. S más C el costo de escasez sobre costo de escasez más costo de excedente, puedes pensarlo incluso en la comida, la comida que tenemos, por ejemplo, que habla de pollo cocido, esa que encuentras en los bufés, en las tiendas de abarrotes o en el mercado libre o las verduras, son de temporada también porque tienen una duración determinada y hasta que podamos hablar del puesto de comida donde la comida es perecedera.

Esto también es estacional y se llama modelo de vendedor de noticias o modelo de vendedor de periódicos, porque es como el vendedor de periódicos que va con el periódico del día y lo distribuyen por la cuadra o por el vecindario y podría terminar con muchos periódicos que hizo, no vender y luego hay que tirarlos o tendrá clientes en un periódico, pero no los tiene y no pueden traerlo mañana porque serán todas noticias. Así que esta es la idea de que la magia de esta fórmula, el problema crítico, y se usa ampliamente en minorista, especialmente para los productos de temporada, también observa otras suposiciones cuando podemos descontar el producto o tener Selvidge o cuando tenemos una penalización por este producto, entonces, ¿qué cambiará y el riesgo crítico? También veremos esto.

**Productos de Temporada:**

Los productos de temporada están en todas partes.
1. Compra única.
2. No podemos reponer nuevamente durante el consumo por parte de los clientes.
3. Se basa en la rentabilidad proveniente de los clientes.

**Supuestos:**

- Se desconoce la demanda.
- Se define el horizonte temporal.
- Las ventas se pierden si no tenemos existencias.
- La Demanda es aleatorio.
- El horizonte de planificación es un período.
- Los productos se volverán obsoletos si tengo un exceso de inventario.

**¿Cómo debo configurar la cantidad?**

- ¿Qué tan costoso será perder un producto?.
- ¿Qué tan costoso será tener un exceso de producto?.

**¿Por qué tenemos períodos únicos?:**

- El plazo de entrega puede ser muy largo.
- Los productos son una edición limitada.
- Los productos ya no se pueden utilizar si se supera el horizonte temporal de las ventas.
- El producto tendrá que venderse con descuento o recuperarse si pasa el horizonte de tiempo.

### Critical Ratio (Proporción Crítica):

**¿Qué está pasando realmente?**

Entonces, ¿qué está sucediendo realmente en el tema crítico? ¿Por qué los cambios raciales a lo largo de la distribución, a veces más que el promedio, a veces menos que el promedio?

Bueno, lo que está sucediendo es que el problema crítico es decirme el riesgo asociado con tener demasiado inventario y el riesgo asociado con tener una pérdida de inventario, por ejemplo, si cuesta demasiado no tener el producto de mi cliente, entonces el tema crítico de decirme, está bien, muévase más allá del promedio, tenga más porque le costará demasiado.

Pero si el problema crítico, si el costo de tener demasiadas existencias es caro y más que el costo de perder al cliente, entonces el problema crítico es moverse un poco en el lado bajo, por debajo del promedio, no más allá del promedio. Entonces, esta es la belleza del problema crítico y puede agregar muchos parámetros, como el rescate, como la penalización, etc., que le identifican el riesgo de tener demasiado o el riesgo de tener muy poco.

Y además, podemos hacer esto para todas nuestras habilidades e incluso podemos integrar el nivel de servicio o los no asegurados. ¿Cuánto me costará?

¿Cuál será el escenario si compro esta cantidad? ¿Cuántas unidades? Se espera que sea corto y lo haremos en inventarios y también me dirá cuántas unidades se espera vender y cuántas unidades se espera que venda con un descuento. También haremos esto en inventarios.

Esto es muy útil y se aplica a gran escala en el retail, en todos los productos que son de temporada.

In [1]:
import pandas as pd
import inventorize as inv
import numpy as np
import array
import matplotlib.pyplot as plt
import os

In [3]:
ruta = os.chdir("C:/ENRIQUE/PROGRAMACIÓN/PROYECTOS/Data Science and Supply Chain with Python/Seccion 18")

In [4]:
mean = 50
sd = 10
c = 1.2
Salvage = 0.7
penality_term = 0.4
price = 3
c = 1.2 # Costo

In [5]:
# Max. Profit
inv.CriticalRatio(price, c, 0, 0) # En este caso el Salvage y penalty = 0

0.6

In [6]:
# Critical Ratio:
(price-c)/(price-c+c)

0.6

In [7]:
inv.MPN_singleperiod(mean, sd, price, c, 0, 0)

{'quantity': 52.533471031358,
 'demand': 50,
 'sd': 10,
 'unitshort': 2.8500369224254065,
 'shortagecost': 8.55011076727622,
 'cost': 63.040165237629594,
 'revenue': 150,
 'profit': 78.40972399509418,
 'soldatfullprice': 47.1499630775746,
 'sold_discount': 5.3835079537834005}

In [8]:
inv.MPN_singleperiod(mean, sd, price, c, Salvage, 0)

{'quantity': 57.81033811522709,
 'demand': 50,
 'sd': 10,
 'unitshort': 1.2427768964900099,
 'shortagecost': 2.8583868619270225,
 'cost': 28.905169057613545,
 'revenue': 150,
 'profit': 83.23644408045942,
 'soldatfullprice': 48.75722310350999,
 'sold_discount': 9.053115011717097}

In [9]:
inv.MPN_singleperiod(mean, sd, price, c, Salvage, penality_term) # Inventario de MPN

{'quantity': 58.957798188835895,
 'demand': 50,
 'sd': 10,
 'unitshort': 1.012102766308183,
 'shortagecost': 2.732677469032094,
 'cost': 29.478899094417947,
 'revenue': 150,
 'profit': 82.78842343654995,
 'soldatfullprice': 48.98789723369182,
 'sold_discount': 9.969900955144077}

In [10]:
inv.EPN_singleperiod(70, 40, 10, price, c, 0, 0)

{'quantity': 70,
 'demand': 40,
 'sd': 10,
 'unitshort': 0.0038215431704769712,
 'shortagecost': 0.011464629511430914,
 'cost': 84.0,
 'revenue': 120,
 'profit': 35.98853537048857,
 'soldatfullprice': 39.996178456829526,
 'sold_discount': 30.003821543170474}

### Preparando los datos para MPN:

In [11]:
retail_clean = pd.read_csv('retail_clean.csv')
retail_clean.head()

Unnamed: 0  Invoice StockCode                          Description  \
0           0   489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS   
1           1   489434    79323P                   PINK CHERRY LIGHTS   
2           2   489434    79323W                  WHITE CHERRY LIGHTS   
3           3   489434     22041         RECORD FRAME 7" SINGLE SIZE    
4           4   489434     21232       STRAWBERRY CERAMIC TRINKET BOX   

   Quantity          InvoiceDate  Price  Customer ID         Country  Revenue  
0        12  2009-12-01 07:45:00   6.95      13085.0  United Kingdom     83.4  
1        12  2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0  
2        12  2009-12-01 07:45:00   6.75      13085.0  United Kingdom     81.0  
3        48  2009-12-01 07:45:00   2.10      13085.0  United Kingdom    100.8  
4        24  2009-12-01 07:45:00   1.25      13085.0  United Kingdom     30.0

In [12]:
retail_clean.columns

Index(['Unnamed: 0', 'Invoice', 'StockCode', 'Description', 'Quantity',
       'InvoiceDate', 'Price', 'Customer ID', 'Country', 'Revenue'],
      dtype='object')

In [13]:
retail_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 779495 entries, 0 to 779494
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Unnamed: 0   779495 non-null  int64  
 1   Invoice      779495 non-null  int64  
 2   StockCode    779495 non-null  object 
 3   Description  779495 non-null  object 
 4   Quantity     779495 non-null  int64  
 5   InvoiceDate  779495 non-null  object 
 6   Price        779495 non-null  float64
 7   Customer ID  779495 non-null  float64
 8   Country      779495 non-null  object 
 9   Revenue      779495 non-null  float64
dtypes: float64(3), int64(3), object(4)
memory usage: 59.5+ MB


In [14]:
retail_clean['InvoiceDate'] = pd.to_datetime(retail_clean['InvoiceDate'])

In [15]:
retail_clean['date'] = retail_clean['InvoiceDate'].dt.strftime('%Y-%m-%d')

In [16]:
retail_clean['date'] = pd.to_datetime(retail_clean['date'])

In [17]:
retail_clean['year'] = retail_clean['date'].dt.year

In [19]:
years_2 = retail_clean[retail_clean.year.isin([2010,2011])]
years_2

Unnamed: 0  Invoice StockCode                      Description  \
30279        45228   493410   TEST001          This is a test product.   
30280        45230   493412   TEST001          This is a test product.   
30281        45234   493414     21844                   RETRO SPOT MUG   
30282        45235   493414     21533        RETRO SPOT LARGE MILK JUG   
30283        45236   493414     37508  NEW ENGLAND CERAMIC CAKE SERVER   
...            ...      ...       ...                              ...   
779490     1067366   581587     22899     CHILDREN'S APRON DOLLY GIRL    
779491     1067367   581587     23254    CHILDRENS CUTLERY DOLLY GIRL    
779492     1067368   581587     23255  CHILDRENS CUTLERY CIRCUS PARADE   
779493     1067369   581587     22138    BAKING SET 9 PIECE RETROSPOT    
779494     1067370   581587      POST                          POSTAGE   

        Quantity         InvoiceDate  Price  Customer ID         Country  \
30279          5 2010-01-04 09:24:00   4.50      12346.0  United Kingdom   
30280          5 2010-01-04 09:53:00   4.50      12346.0  United Kingdom   
30281         36 2010-01-04 10:28:00   2.55      14590.0  United Kingdom   
30282         12 2010-01-04 10:28:00   4.25      14590.0  United Kingdom   
30283          2 2010-01-04 10:28:00   2.55      14590.0  United Kingdom   
...          ...                 ...    ...          ...             ...   
779490         6 2011-12-09 12:50:00   2.10      12680.0          France   
779491         4 2011-12-09 12:50:00   4.15      12680.0          France   
779492         4 2011-12-09 12:50:00   4.15      12680.0          France   
779493         3 2011-12-09 12:50:00   4.95      12680.0          France   
779494         1 2011-12-09 12:50:00  18.00      12680.0          France   

        Revenue       date  year  
30279     22.50 2010-01-04  2010  
30280     22.50 2010-01-04  2010  
30281     91.80 2010-01-04  2010  
30282     51.00 2010-01-04  2010  
30283      5.10 2010-01-04  2010  
...         ...        ...   ...  
779490    12.60 2011-12-09  2011  
779491    16.60 2011-12-09  2011  
779492    16.60 2011-12-09  2011  
779493    14.85 2011-12-09  2011  
779494    18.00 2011-12-09  2011  

[749216 rows x 12 columns]

In [20]:
total = years_2.groupby(['year','Description']).agg(total_sales = ('Quantity',np.sum),
                                                    price = ('Price','mean')).reset_index()
total

year                          Description  total_sales      price
0     2010    DOORMAT UNION JACK GUNS AND ROSES          167   6.640196
1     2010             3 STRIPEY MICE FELTCRAFT          399   1.941429
2     2010        4 PURPLE FLOCK DINNER CANDLES          200   2.150000
3     2010                      ANIMAL STICKERS          169   0.210000
4     2010          BLACK PIRATE TREASURE CHEST            6   1.650000
...    ...                                  ...          ...        ...
8102  2011      ZINC T-LIGHT HOLDER STARS SMALL         4894   0.836975
8103  2011       ZINC TOP  2 DOOR WOODEN SHELF             9  16.950000
8104  2011     ZINC WILLIE WINKIE  CANDLE STICK         2432   0.873833
8105  2011          ZINC WIRE KITCHEN ORGANISER           25   7.175000
8106  2011     ZINC WIRE SWEETHEART LETTER TRAY           80   3.454000

[8107 rows x 4 columns]

In [21]:
expected = total.groupby('Description').agg(expected_Demand= ('total_sales',np.mean),
                                            sd= ('total_sales','std'),
                                            price=('price',np.mean)).reset_index()
expected

Description  expected_Demand          sd  \
0       DOORMAT UNION JACK GUNS AND ROSES            167.0         NaN   
1                3 STRIPEY MICE FELTCRAFT            399.0         NaN   
2           4 PURPLE FLOCK DINNER CANDLES            162.0   53.740115   
3           50'S CHRISTMAS GIFT BAG LARGE           1885.0         NaN   
4                         ANIMAL STICKERS            169.0         NaN   
...                                   ...              ...         ...   
5146      ZINC T-LIGHT HOLDER STARS SMALL           4894.0         NaN   
5147       ZINC TOP  2 DOOR WOODEN SHELF              69.0   84.852814   
5148     ZINC WILLIE WINKIE  CANDLE STICK           2818.0  545.886435   
5149          ZINC WIRE KITCHEN ORGANISER             25.0         NaN   
5150     ZINC WIRE SWEETHEART LETTER TRAY             80.0         NaN   

          price  
0      6.640196  
1      1.941429  
2      2.224286  
3      1.248073  
4      0.210000  
...         ...  
5146   0.836975  
5147  14.753571  
5148   0.860579  
5149   7.175000  
5150   3.454000  

[5151 rows x 4 columns]

In [22]:
expected['sd']

0              NaN
1              NaN
2        53.740115
3              NaN
4              NaN
           ...    
5146           NaN
5147     84.852814
5148    545.886435
5149           NaN
5150           NaN
Name: sd, Length: 5151, dtype: float64

### Creando un margen de error:

In [23]:
def margin_error(dataframe):
    if(pd.isna(dataframe['sd'])):
        a= dataframe['expected_Demand']*0.1
    else:
        a= dataframe['sd']
    return a

In [24]:
expected['sd1'] = expected.apply(margin_error,axis= 1)
expected['sd1']

0        16.700000
1        39.900000
2        53.740115
3       188.500000
4        16.900000
           ...    
5146    489.400000
5147     84.852814
5148    545.886435
5149      2.500000
5150      8.000000
Name: sd1, Length: 5151, dtype: float64

In [25]:
expected['cost'] = expected['price']*0.4 # EL 40% de la demanda esperada
expected['cost']

0       2.656078
1       0.776571
2       0.889714
3       0.499229
4       0.084000
          ...   
5146    0.334790
5147    5.901429
5148    0.344232
5149    2.870000
5150    1.381600
Name: cost, Length: 5151, dtype: float64

In [26]:
empty_data = pd.DataFrame()

### Aplicar MPN a todos los datos:

In [27]:
for i in range(expected.shape[0]):
    a= inv.MPN_singleperiod(expected.loc[i,'expected_Demand'],
                             expected.loc[i,'sd1'],
                             expected.loc[i,'price'], expected.loc[i,'cost'], 0, 0)
    b= pd.DataFrame(a,index=[0])
    b['description']=expected.loc[i,'Description']
    empty_data= pd.concat([empty_data,b],axis=0)
     
empty_data

C:\Users\enriq\anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:2128: RuntimeWarning: invalid value encountered in multiply
  lower_bound = _a * scale + loc
C:\Users\enriq\anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:2129: RuntimeWarning: invalid value encountered in multiply
  upper_bound = _b * scale + loc


quantity  demand          sd   unitshort  shortagecost         cost  \
0    171.230897   167.0   16.700000    4.759562     31.604423   454.802691   
0    409.108549   399.0   39.900000   11.371647     22.077241   317.702011   
0    175.614903   162.0   53.740115   15.316131     34.067452   156.247088   
0   1932.755929  1885.0  188.500000   53.723196     67.050492   964.888501   
0    173.281566   169.0   16.900000    4.816562      1.011478    14.555652   
..          ...     ...         ...         ...           ...          ...   
0   5017.988072  4894.0  489.400000  139.480807    116.741919  1679.971805   
0     90.497215    69.0   84.852814   24.183365    356.791006   534.062848   
0   2956.298747  2818.0  545.886435  155.579650    133.888613  1017.651706   
0     25.633368    25.0    2.500000    0.712509      5.112254    73.567765   
0     82.026777    80.0    8.000000    2.280030      7.875222   113.328195   

        revenue       profit  soldatfullprice  sold_discount  \
0   1108.912745   622.505631       162.240438       8.990458   
0    774.630000   434.850748       387.628353      21.480197   
0    360.334286   170.019746       146.683869      28.931034   
0   2352.618349  1320.679356      1831.276804     101.479125   
0     35.490000    19.922870       164.183438       9.098128   
..          ...          ...              ...            ...   
0   4096.154622  2299.440898      4754.519193     263.468879   
0   1017.996429   127.142575        44.816635      45.680580   
0   2425.112237  1273.571918      2662.420350     293.878397   
0    179.375000   100.694981        24.287491       1.345877   
0    276.320000   155.116583        77.719970       4.306806   

                            description  
0     DOORMAT UNION JACK GUNS AND ROSES  
0              3 STRIPEY MICE FELTCRAFT  
0         4 PURPLE FLOCK DINNER CANDLES  
0         50'S CHRISTMAS GIFT BAG LARGE  
0                       ANIMAL STICKERS  
..                                  ...  
0       ZINC T-LIGHT HOLDER STARS SMALL  
0        ZINC TOP  2 DOOR WOODEN SHELF   
0      ZINC WILLIE WINKIE  CANDLE STICK  
0           ZINC WIRE KITCHEN ORGANISER  
0      ZINC WIRE SWEETHEART LETTER TRAY  

[5151 rows x 11 columns]

In [28]:
empty_data.iloc[1,:]

quantity                             409.109
demand                                   399
sd                                      39.9
unitshort                            11.3716
shortagecost                         22.0772
cost                                 317.702
revenue                               774.63
profit                               434.851
soldatfullprice                      387.628
sold_discount                        21.4802
description         3 STRIPEY MICE FELTCRAFT
Name: 0, dtype: object

### Conclusión:

Todos, así que hemos hecho el bucle for sobre los datos vacíos para obtener una descripción de Pritam y ahora veamos anoche, un bistec, por ejemplo.

El primero. Miro entonces la cantidad 409, esa fue la cantidad máxima que propuso la categorización, mientras que la demanda es 399, la desviación estándar es treinta y nueve coma nueve y pueden ver eso.

El costo es de 317 y solo hay 774 con fines de lucro, venderán 387 a precio completo y nosotros venderemos 21.48 con descuento. Entonces, ¿por qué la cantidad se volvió menor que la demanda? Porque para este producto específico el costo de almacenamiento es más caro que el costo de no vender tanto en el costo de perder al cliente, el costo de almacenar el producto es más caro que el costo de perder al cliente. Si el costo de perder al cliente fue más caro, encontrará la cantidad más que eso, quiso decir, de hecho, podemos intentar ahora cambiar el costo a cero puntos y luego volver a intentarlo.

Y ahora ve que la cantidad se convirtió en más que la demanda porque ahora el costo de perder al cliente o perder la ganancia o el costo de oportunidad es mucho mayor que el costo del transporte por camión. Entonces, el problema crítico es decirnos, mire, necesita comprar más para esto porque perder al cliente es costoso para usted.

Entonces, debido a que aquí nos dimos cuenta del costo de un rollo del 40%, encontrará que para todos los datos en esta cantidad será más que la demanda esperada, pero en el caso de que tengamos datos que con cada uno con un costo específico y un precio específico , entonces la cantidad variará según la variación entre el precio y el costo. Tenga en cuenta que hacemos esto para los productos de temporada, los productos que compramos en Atlanta como una compra única cada temporada y podemos tener, como dijimos, muchos ejemplos como la tienda de comestibles, como los modelos de moda, las colecciones de moda, muchas de las cosas que compramos una vez se pueden aplicar al modelo de proveedor más nuevo o de temporada.

### Resumen de Inventario Estacional:

Así que ahora recapitulemos esta sección larga pero gratificante con respecto a los productos de temporada, dijimos que los productos de temporada solo duran una temporada después de este año podrían venderlos con un rescate o podrían venderlos con un descuento y el problema crítico es que C es sobre C como más C y E, podemos tener algunas variaciones.

Por qué salvar lo que llamamos CR = Cs / Cs + Ce, y menos el salvamento, porque el salvamento es dinero que me regresa si tienes una penalización. Estas penalizaciones terminan aquí, así que sí, 

- CRsalvamento = Cs / Cs + Ce - s (salvamento)
- CR = Cs + B / Cs + B + Ce - s

Ya que algunas de estas son las variaciones del rescate de la relación crítica y agrega la penalización cuando cree que podría perder al cliente, por ejemplo, y quiere sopesar más el riesgo de no tener el producto a mano, es muy útil para productos de temporada.

Hemos aplicado muchas funciones, como:
- MPN para el máximo beneficio, normal, 
- MPP para el máximo beneficio para todos nosotros, y hemos aplicado 
- EPN y EPPY y hemos aplicado también 
- El problema crítico (CR) 

Y todo esto te dirá exactamente cómo cuánto se espera que obtenga como ganancia y cuál es su costo esperado, cuánto venderá con un descuento, cuánto venderá al precio completo, todas estas funciones ¿por qué?.

Es importante saber si es MPN, porque como sabemos antes, caracterizar la distribución es la parte más importante. Antes de continuar con el análisis para saber qué distribución se ajusta mejor tu demanda, esta sección te aconsejo que la revises nuevamente y trabajes en los ejercicios, en las asignaciones, y tal vez practiques con algún dato que tengas de tu vida real o su entorno de práctica real en su industria y estoy seguro de que encontrará esto útil en futuras decisiones que tome en su Cadena de Suministro.